### DISCUSSION TITLE
Report error running "Multi-Timescale" examples

### DISCUSSION THREAD

**Discussion by xiaocaoren** [neuralhydrology/neuralhydrology#140] [Q&A] [answered]
I am a newer for neuralhydrology. I downloaded the datasets and began to run the examples for learn. I can go throught the example of "01-Introduction” in Jupyter. But when I got stunk in "04-Multi-Timescale". It reported the errors when I run the following code:
```
# by default we assume that you have at least one CUDA-capable NVIDIA GPU
if torch.cuda.is_available():
    start_run(config_file=Path("1_basin.yml"))

# fall back to CPU-only mode
else:
    start_run(config_file=Path("1_basin.yml"), gpu=-1)
```

And the errors are as follows:

> 

> TypeError                                 Traceback (most recent call last)
> Cell In[2], line 3
>       1 # by default we assume that you have at least one CUDA-capable NVIDIA GPU
>       2 if torch.cuda.is_available():
> ----> 3     start_run(config_file=Path("1_basin.yml"))
>       5 # fall back to CPU-only mode
>       6 else:
>       7     start_run(config_file=Path("1_basin.yml"), gpu=-1)
> 
> File f:\workspace\neuralhydro\neuralhydrology\neuralhydrology\nh_run.py:76, in start_run(config_file, gpu)
>      73 if gpu is not None and gpu < 0:
>      74     config.device = "cpu"
> ---> 76 start_training(config)
> 
> File f:\workspace\neuralhydro\neuralhydrology\neuralhydrology\training\train.py:20, in start_training(cfg)
>      18     raise ValueError(f"Unknown head {cfg.head}.")
>      19 trainer.initialize_training()
> ---> 20 trainer.train_and_validate()
> 
> File f:\workspace\neuralhydro\neuralhydrology\neuralhydrology\training\basetrainer.py:214, in BaseTrainer.train_and_validate(self)
>     211     for param_group in self.optimizer.param_groups:
>     212         param_group["lr"] = self.cfg.learning_rate[epoch]
> --> 214 self._train_epoch(epoch=epoch)
>     215 avg_losses = self.experiment_logger.summarise()
>     216 loss_str = ", ".join(f"{k}: {v:.5f}" for k, v in avg_losses.items())
> 
> File f:\workspace\neuralhydro\neuralhydrology\neuralhydrology\training\basetrainer.py:283, in BaseTrainer._train_epoch(self, epoch)
>     281 # Iterate in batches over training set
>     282 nan_count = 0
> --> 283 for data in pbar:
>     285     for key in data.keys():
>     286         if not key.startswith('date'):
> 
> File D:\Users\DELL\anaconda3\envs\neuralhydrology\lib\site-packages\tqdm\std.py:1178, in tqdm.__iter__(self)
>    1175 time = self._time
>    1177 try:
> -> 1178     for obj in iterable:
>    1179         yield obj
>    1180         # Update and possibly print the progressbar.
>    1181         # Note: does not call self.update(1) for speed optimisation.
> 
> File D:\Users\DELL\anaconda3\envs\neuralhydrology\lib\site-packages\torch\utils\data\dataloader.py:441, in DataLoader.__iter__(self)
>     439     return self._iterator
>     440 else:
> --> 441     return self._get_iterator()
> 
> File D:\Users\DELL\anaconda3\envs\neuralhydrology\lib\site-packages\torch\utils\data\dataloader.py:388, in DataLoader._get_iterator(self)
>     386 else:
>     387     self.check_worker_number_rationality()
> --> 388     return _MultiProcessingDataLoaderIter(self)
> 
> File D:\Users\DELL\anaconda3\envs\neuralhydrology\lib\site-packages\torch\utils\data\dataloader.py:1042, in _MultiProcessingDataLoaderIter.__init__(self, loader)
>    1035 w.daemon = True
>    1036 # NB: Process.start() actually take some time as it needs to
>    1037 #     start a process and pass the arguments over via a pipe.
>    1038 #     Therefore, we only add a worker to self._workers list after
>    1039 #     it started, so that we do not call .join() if program dies
>    1040 #     before it starts, and __del__ tries to join but will get:
>    1041 #     AssertionError: can only join a started process.
> -> 1042 w.start()
>    1043 self._index_queues.append(index_queue)
>    1044 self._workers.append(w)
> 
> File D:\Users\DELL\anaconda3\envs\neuralhydrology\lib\multiprocessing\process.py:121, in BaseProcess.start(self)
>     118 assert not _current_process._config.get('daemon'), \
>     119        'daemonic processes are not allowed to have children'
>     120 _cleanup()
> --> 121 self._popen = self._Popen(self)
>     122 self._sentinel = self._popen.sentinel
>     123 # Avoid a refcycle if the target function holds an indirect
>     124 # reference to the process object (see bpo-30775)
> 
> File D:\Users\DELL\anaconda3\envs\neuralhydrology\lib\multiprocessing\context.py:224, in Process._Popen(process_obj)
>     222 @staticmethod
>     223 def _Popen(process_obj):
> --> 224     return _default_context.get_context().Process._Popen(process_obj)
> 
> File D:\Users\DELL\anaconda3\envs\neuralhydrology\lib\multiprocessing\context.py:336, in SpawnProcess._Popen(process_obj)
>     333 @staticmethod
>     334 def _Popen(process_obj):
>     335     from .popen_spawn_win32 import Popen
> --> 336     return Popen(process_obj)
> 
> File D:\Users\DELL\anaconda3\envs\neuralhydrology\lib\multiprocessing\popen_spawn_win32.py:93, in Popen.__init__(self, process_obj)
>      91 try:
>      92     reduction.dump(prep_data, to_child)
> ---> 93     reduction.dump(process_obj, to_child)
>      94 finally:
>      95     set_spawning_popen(None)
> 
> File D:\Users\DELL\anaconda3\envs\neuralhydrology\lib\multiprocessing\reduction.py:60, in dump(obj, file, protocol)
>      58 def dump(obj, file, protocol=None):
>      59     '''Replacement for pickle.dump() using ForkingPickler.'''
> ---> 60     ForkingPickler(file, protocol).dump(obj)
> 
> TypeError: cannot pickle '_thread.lock' object
> 

I have transferred to another computer with the same code, and I debugged it in IDE environments, like Pycharm and Spyder, neither worked. Could anybody encountered the similar situation?

Thank you so much！

**Comment by gauchm** [comment, marked as answer]
This is likely an issue with the conda environment and version mismatches. See previous cases where this happened: #75, #81. My suggestion would be to create a fresh conda environment from the provided [environment yaml files](https://github.com/neuralhydrology/neuralhydrology/tree/master/environments).

If all else fails, setting num_workers to 0 should be a workaround.

**Reply by xiaocaoren** [reply]
Yes, you are correct. It is awesome that everything go well when I change the num_workers into 0. And the program run much faster than ever. I looked through the previous cases and probobaly got the point. I guess it was because I run in a single computer and there is no parallel processing. 
Thank you very much.

**Comment by mengziwen2000** [comment]
In Tutorial 00 - Data-Prerequisites, when downloading hourly data, it is recommended to download netCDF files because netCDF files load faster than csv files. But in Tutorial 04 - Multi-Timescale tutorial, ```yml.forcings``` are ```nldas_hourly``` and ```daymet```, and ```nldas_hourly``` is the folder containing the csv file. So how do I modify ```yml.forcings``` so as to call the netCDF file?

![image](https://github.com/user-attachments/assets/3806a395-9e1f-4b1a-b965-02dd411848db)
Tutorial 04 - Multi-Timescale Prediction

**Reply by gauchm** [reply]
The code will first try loading the nc file, and only if it doesn't exist use the csv files. See https://github.com/neuralhydrology/neuralhydrology/blob/master/neuralhydrology/datasetzoo/hourlycamelsus.py#L269.

You'd see a warning in the logs if the nc is not used: https://github.com/neuralhydrology/neuralhydrology/blob/master/neuralhydrology/datasetzoo/hourlycamelsus.py#L149.

**Reply by mengziwen2000** [reply]
Thanks. I am clear now. ```yml.forcings``` are not to be changed. When the nc file exists in the hourly folder, the model calls the nc file directly.

**Comment by mengziwen2000** [comment]
As I view the structural diagram of the MTS-LSTM model, I think that the training inputs for the model i.e. day-scale data and hour-scale data will be different in terms of input length. As a simple example, it is possible to train the MTS-LSTM model with daily scale data for the period from 2010-01-01 to 2011-01-01 and hourly scale data for the period of a few days between 2010-12-20 and 2011-01-01. However, when I look at the papers related to MTS-LSTM models and the ```.yml``` files, I find that the training period and the testing period are the same for both scales of data.
I now have ten years of continuous daily scale data as well as intermittent (flood season only) hourly scale data, so is it not possible to get hourly scale flood forecasts with the MTS-LSTM model because the length of the data is not the same for both scales.

**Reply by gauchm** [reply]
You could train an MTS-LSTM where you calculate the loss on both the daily and hourly outputs if ground truth for both is available, and only on the daily outputs if only daily ground truth is available. I haven't tried this, but I would be confident that if you have sufficient hourly labels this could work pretty well.
